### PR TITLE
(PUP-7946) Add Puppet::Pal, API to run scripts

### PR DESCRIPTION
**Pull request by hlindberg**
This makes an API available that shows only a small surface area to "all
things puppet". In this version it is possible to run a string like in
the style of puppet apply -e, or a manifest (.pp) file.

There is more to come in this API.

### CODE REVIEW

File: lib/puppet_pal.rb [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  # @param settings_hash [Hash] a hash of settings - currently not used for anything, defaults to empty hash
+  # @return [Object] returns what the given block returns
+  # @yieldparam [Puppet::Pal] context, a context that responds to Puppet::Pal methods
+  #
+  def self.in_tmp_environment(env_name,
+      modulepath:    [],
+      settings_hash: {}
+    )
+    unless env_name.is_a?(String) && env_name.length > 0
+      raise ArgumentError(_("Puppet Pal: temporary environment name must be a non empty string, got '%{env_name}'") % {:env_name => env_name})
```
Comment (pcarlisle) [inline lib/puppet_pal.rb]: this and the next two argumenterrors are wrong syntax

**Comment by pcarlisle** [comment]
I'm not sure in_tmp_environment is working right. When I step through with a debugger I don't see the temporary environment actually get used in `main`, it just has *root* for everything with no modulepath.

rough integration at https://github.com/pcarlisle/bolt/tree/integrate-puppet

Looking more at it, I'm not sure how to pass in the environment directory, which should be needed to look up a plan by name.

Should I have to initialize settings for this to work? I know we discussed this a little. It seems heavy handed but I'm not sure how to avoid it.

**Comment by pcarlisle** [comment]
I'm also not sure why it looks up a node and gets facts.

File: lib/puppet_pal.rb (hunk 1 above)
Comment (hlindberg) [reply lib/puppet_pal.rb]: Fixed.

**Comment by hlindberg** [comment]
It gets facts in order to let a plan decide on how to run local tasks. That is to get $facts populated when evaluating. I plan to change this so it can be set for a group of operations - now it gets the fact for each call.

**Comment by hlindberg** [comment]
The call to set up an existing environment (i.e. a directory that has environment layout) is not there yet. It should be able to find plans and tasks from modules on the modulepath with the current API. (But, as you found, that may not actually work yet - I have more tests to write to assert that).

**Comment by hlindberg** [comment]
If settings needs to be initiated, that should be done "by the API" as it is a consequence of how puppet is implemented.

**Pull request merged**